### ISSUE TITLE
source mode error "Loading input failed: loading package failed"

### ISSUE THREAD

**Issue by rgarcia** [golang/mock#283]
The commit to switch to go modules (837231f7bb377b365da147e5ff6c031b12f0dfaa) leads to `mockgen` producing a "Loading input failed" error. I've confirmed that it works with the previous commit, bfa1e97ed7a159b123865559eabdec69b78fe1d5. Repro script: https://gist.github.com/3d0e2b1e77a636851c1f51c10a9a15ef

Go version is `go version go1.10 linux/amd64`.

**Comment by rgarcia** [comment]
tested using `go version go1.11.1 darwin/amd64` and it also produces the error

**Comment by bithavoc** [comment]
I'm not able to use this package with Go modules on.

I've tried:

```
export GOFLAGS=-mod=vendor
export GO111MODULE=on
```

Any suggestions?

**Comment by bithavoc** [comment]
Basically what happens is that when mockgen is executed in a project outside `GOPATH` it generates:

```go
import (
	x "."
	context "context"
	gomock "github.com/golang/mock/gomock"
	reflect "reflect"
)
```

Then the compiler complains:

```
 local import "." in non-local package
```

**Comment by bithavoc** [comment]
Oops, sorry, I'm new to go modules. Once I executed the following commands outside my go module-enabled project, it picked the updated `mockgen` binary with the fix:

```
go install github.com/golang/mock/mockgen
```

**Comment by poy** [comment]
@rgarcia Thanks for reporting this. I'll investigate.

**Comment by poy** [comment]
This seems to be an artifact of [golang.org/x/tools/go/packages](https://godoc.org/golang.org/x/tools/go/packages) having more strict rules about where the `.go` files are stored.

In the case of your linked script `interface.go` is not in a proper location (e.g., not in a repo with `go.mod` or even in a GOPATH). Therefore, `packages.Load` can't find a package (as its not in one).

I updated the script:
```bash
#!/usr/bin/env bash

set -ex

cd /tmp
rm -rf /tmp/mock
mkdir -p /tmp/mock/src/github.com/golang
cd /tmp/mock
export GOPATH=$PWD

pushd src/github.com/golang
    git clone https://github.com/golang/mock
    (cd mock
         git checkout bfa1e97ed7a159b123865559eabdec69b78fe1d5 &&
         go build -o /tmp/mock/mockgen-before ./mockgen)
    rm -rf mock && git clone https://github.com/golang/mock
    (cd mock &&
         git checkout 837231f7bb377b365da147e5ff6c031b12f0dfaa &&
         go build -o /tmp/mock/mockgen-after ./mockgen)
popd

mkdir -p $GOPATH/src/github.com/bar/foo

cat <<EOF >$GOPATH/src/github.com/bar/foo/interface.go
package foo

type Interface interface {
	Bar()
}
EOF

# works
./mockgen-before -package foo -destination interface_mock.go -source $GOPATH/src/github.com/bar/foo/interface.go Interface

# NOW works
if [ "$GO111MODULE" == "on" ]; then
    echo "module github.com/bar/foo" > $GOPATH/src/github.com/bar/foo/go.mod
    pushd $GOPATH/src/github.com/bar/foo/
        /tmp/mock/mockgen-after -package foo -destination interface_mock.go -source interface.go Interface
    popd
else
    ./mockgen-after -package foo -destination interface_mock.go -source $GOPATH/src/github.com/bar/foo/interface.go Interface
fi
```

**Comment by rgarcia** [comment]
@poy thanks for figuring out the cause! I'm fine closing this for now.

Now that I dig into this I think the error message that the script generates is slightly different than the actual error we're seeing at Clever. We build mockgen from a vendor/'d copy pulled down by `dep`. `dep` prunes the newly introduced `vendor/` folder, and as far as I can tell does not interpret `go.mod`. This means that building from `vendor/github.com/golang/mock/mockgen` yields something slightly different than what gets built if you clone the repo. I think this is an issue with `dep` and how we are building mockgen, not with mockgen itself. Thanks for your help.

**Comment by undercover87** [comment]
Hello, I'm getting the same error message. To my understanding, this is a bug and still open .. right?
Were the changes on the OP's script proposed above a sort of workaround?

**Comment by balshetzer** [comment]
This problem as experienced by other users turned out to be an error in
their configuration exposed by mockgen becoming more strict about how it
loads files.

Please give us the details of the problem you're experiencing and we can
see what the cause is and whether it requires a fix in mockgen or I. Your
code.

**Comment by undercover87** [comment]
Hello, 
sorry for not posting details, I was confused on what to actually post. 
It turns out it was a configuration issue for me as well, GOPATH (and GOROOT) had not been set properly.

In any case, thanks for the prompt response!

**Comment by prasoontelang** [comment]
I was experiencing this issue, I have made a small example in the repo below:

https://github.com/prasoontelang/mockgen-283

Please let me know if I have missed anything. Thanks!

**Comment by wanfintara** [comment]
Hi, currently I'm getting this error. After execute `go mod vendor` then `go generate ./...`

got this
```
can't load package: package github.com/golang/mock/mockgen: cannot find package "." in:
        *myrepo*/vendor/github.com/golang/mock/mockgen
```

but if I delete the `vendor` folder everything works. Do i need run above script sir ??

**Comment by Xanvial** [comment]
there's a workaround of adding import to mockgen (or other requirement like model), so it will be added to vendor folder. for example adding this to one of test file 
```
import (
	_ "github.com/golang/mock/mockgen/model"
)
```

**Comment by Higan** [comment]
I just encountered exactly the same error message, and it turned out to be an import cycle issue.

**Comment by valters** [comment]
Please also see fixes and workarounds in #494
